byron/ledger/executable-spec/test/Ledger/Delegation/Properties.hs

```diff
@@ -276,6 +319,114 @@ dcertsAreTriggered = withTests 300 $ property $
   -- uncover the presence of errors.
   forAll (nonTrivialTrace 1000) >>= dcertsAreTriggeredInTrace
 
+dblockTracesAreClassified :: Property
+dblockTracesAreClassified = withTests 200 $ property $ do
+  let (tl, step) = (1000, 100)
+  tr <- forAll (trace @DBLOCK tl)
+  classifyTraceLength tr tl step
+  -- Classify the traces by the total number of delegation certificates on
+  -- them.
+  classifySize "total dcerts" (traceDCerts tr) length tl step
+  success
+
+-- | Extract the delegation certificates in the blocks, in the order they would
+-- have been applied.
+traceDCerts :: Trace DBLOCK -> [DCert]
+traceDCerts tr = concat $ _blockCerts <$> traceSignals OldestFirst tr
+
+relevantCasesAreCovered :: Property
+relevantCasesAreCovered = withTests 400 $ property $ do
+  let tl = 1000
+  tr <- forAll (trace @DBLOCK tl)
+
+  -- 70% of the traces must contain are as many delegation certificates as
+  -- blocks.
+  cover 70
+        "there are at least as many delegation certificates as blocks"
+        (traceLength tr <= length (traceDCerts tr))
+
+  -- 70% of the traces must contain at most 25% of blocks with empty delegation
+  -- payload.
+  cover 70
+        "at most 25% of the blocks can contain empty delegation payload"
+        (ratio emptyDelegationPayload tr <= 0.25)
+
+  -- 50% of the traces must contain at least 30% of delegations to this epoch.
+  cover 50
+        "at least 30% of the certificates delegate in this epoch"
+        (0.3 <= ratio thisEpochDelegations tr)
+
+  -- 50% of the traces must contain at least 50% of delegations to the next
+  -- epoch.
+  cover 50
+        "at least 50% of the certificates delegate in the next epoch"
+        (0.5 <= ratio nextEpochDelegations tr)
+
+  -- 90% of the traces must contain at least 30% of self-delegations.
+  cover 90
+       "at least 30% of the certificates self delegate"
+       (0.3 <= ratio selfDelegations tr)
+
+  -- 20% of the traces must contain at least 10% of delegations to the same
+  -- delegate.
+  cover 20
+        "at least 10% of the certificates delegate to the same key"
+        (0.1 <= ratio multipleDelegations tr)
+  where
+    ratio :: Integral a
+          => (Trace DBLOCK -> a)
+          -> Trace DBLOCK
+          -> Double
+    ratio f tr = fromIntegral (f tr) / fromIntegral (traceLength tr)
+
+    selfDelegations tr = length
+                       $ filter idDeleg
+                       $ fmap delegatorDelegate (traceDCerts tr)
+      where
+        idDeleg (vks, vk) = owner vks == owner vk
+
+    -- Count the number of delegations to the same key in a given trace.
+    multipleDelegations tr = -- Get the (delegator, delegate) pairs
```

Comment:
For easier reading, I'd appreciate function types for all the functions below.

Reply:
:heavy_plus_sign:

Reply:
Done.